File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

/*
모든 경우의수를 카운팅하는데, 재귀로 풀면될것 같다.
+/- 로 돌리면 될듯 
*/

void DFS(const vector<int> &arr,int target,int idx,int sum)
{
    if(idx+1 == (int)arr.size())
    {
        if(sum == target)  
        {
            answer++;
            return;
        }
        
        else return;
    }
    
    DFS(arr,target,idx+1,sum+arr[idx+1]);
    DFS(arr,target,idx+1,sum-arr[idx+1]);

    return;
}

int solution(vector<int> numbers, int target) 
{
    DFS(numbers,target,-1,0);
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cs
using System;

public class Solution 
{
    private int answer;
    
    private void Dfs(int[] numbers, int target, int sum, int idx)
    {
        if(idx ==numbers.Length)
        {
            if(target == sum) 
            {
                answer++;
            }
            return;
        }
        
        Dfs(numbers,target,sum + numbers[idx],idx+1);
        Dfs(numbers,target,sum - numbers[idx],idx+1);
    }
    
    public int solution(int[] numbers, int target) 
    {
        answer = 0;
        Dfs(numbers,target,0,0);
        return answer;
    }
}